moshi/src/main/java/com/squareup/moshi/EnumJsonAdapter.java

```diff
@@ -0,0 +1,53 @@
+package com.squareup.moshi;
+
+import java.io.IOException;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Type;
+
+/**
+ * Convert Enum types to JSON string values. This uses the enum class's
+ * <pre>toString()</pre> to allow for customization.
+ *
+ * @param <T> Enum class to adapt
+ */
+public class EnumJsonAdapter<T extends Enum<T>> extends JsonAdapter<T> {
+  public static final Factory FACTORY = new Factory() {
+    @Override public JsonAdapter<? extends Enum> create(Type type, AnnotatedElement annotations,
+        Moshi moshi) {
+      Class<?> rawType = Types.getRawType(type);
+      if (Enum.class.isAssignableFrom(rawType)) {
+        //noinspection unchecked
+        return new EnumJsonAdapter<>((Class<? extends Enum>) rawType).nullSafe();
+      }
+      return null;
+    }
+  };
+
+  private final Class<T> clazz;
+
+  public EnumJsonAdapter(Class<T> clazz) {
+    this.clazz = clazz;
+  }
+
+  @Override
+  public T fromJson(JsonReader reader) throws IOException {
+    String name = reader.nextString();
+    try {
+      //noinspection unchecked
+      T[] values = (T[]) clazz.getMethod("values").invoke(null);
```

Comment:
You don't need reflection for this. `class.getEnumConstants()` will give you a `T[]`. This should also only be done once, as it's expensive.